pdr_backend/pdr_dashboard/pages/feeds.py

```diff
@@ -1,22 +1,128 @@
-from dash import html, dcc
+from typing import List, Dict, Any, Union, Tuple
+
+from dash import html, dcc, dash_table
 import dash_bootstrap_components as dbc
+from pdr_backend.pdr_dashboard.dash_components.util import (
+    get_feed_payouts_stats_from_db,
+    get_feed_subscription_stats_from_db,
+    col_to_human,
+    find_with_key_value,
+)
 
 
 class FeedsPage:
+    def __init__(self, app):
+        self.app = app
+
     def layout(self):
         return html.Div(
             [
                 dcc.Store(id="user-payout-stats"),
-                html.H1(
-                    "Feeds page content",
-                    id="page_title_feeds",
-                ),
                 dcc.Loading(
                     id="loading",
                     type="default",
-                    children=[],
+                    children=[
+                        self.get_main_container(),
+                    ],
                     custom_spinner=html.H2(dbc.Spinner(), style={"height": "100%"}),
                 ),
             ],
             style={"height": "100%"},
         )
+
+    def get_main_container(self):
+        feed_stats = get_feed_payouts_stats_from_db(self.app.lake_dir)
+        feed_subscriptions = get_feed_subscription_stats_from_db(
+            self.app.lake_dir, self.app.network_name
+        )
+
+        feed_cols, feed_data = self.get_feeds_data_for_feeds_table(
+            feed_stats, feed_subscriptions
+        )
+
+        return html.Div(
+            [self.get_feeds_table_area(feed_cols, feed_data)],
+            className="main-container",
+        )
+
+    def get_feeds_table_area(
+        self,
+        columns,
+        feeds_data,
+    ):
+        return html.Div(
+            [
+                dash_table.DataTable(
+                    id="feeds_table",
+                    columns=columns,
+                    data=feeds_data,
+                    sort_action="native",  # Enables sorting feature
+                )
+            ],
+            style={"width": "100%"},
+        )
+
+    def get_feeds_stat_with_contract(
+        self, contract: str, feed_stats: List[Dict[str, Any]]
+    ) -> Union[Tuple[float, float, float], None]:
+        result = find_with_key_value(feed_stats, "contract", contract)
```

Comment:
I this particular case, why are we not using a query? Instead of fetching the entire feeds data and filtering it for a contract, wouldn't it make sense to select that feed directly? Maybe I understood the rest wrong, but the "find_with_key_value" function has a weird vibe to me. If it's really needed, I think we already have this logic somewhere with filters in pdr dashboard.

Reply:
Let me summarize what we do here:
1. Retrieve feed data from the prediction table by grouping contracts. * * *
2. Gather payout stats by grouping them with contracts from the payouts table.

Then, we merge them by selecting contracts on line 98 of feeds.py, so all result rows are used in the data:
`result = self.get_feeds_stat_with_contract(feed["contract"], feed_stats)`

I contemplated merging these queries, but there are some drawbacks:
- Firstly, as we already query the database for the feeds on the home screen (the first step that I marked above) , so we do not query again for the feeds page.
- Since there is no contract column on other tables, we split the ID to get the contract address. If we use the "JOIN" statement, it could lead to performance issues.